### PR TITLE
Moved stuff from CoreBundle

### PR DESCRIPTION
**Pull request by core23**
I am targeting this branch, because this is BC.

## Changelog

```markdown
### Added
- Added all doctrine stuff from `SonataCoreBundle`
```

## Subject

This will move all doctrine stuff from the old `SonataCoreBundle` to this bundle (https://github.com/sonata-project/dev-kit/issues/348).

### CODE REVIEW

**Comment by core23** [comment]
Just a sidenote: The classes are non-final, so we can extend this in the core bundle in with minor release

File: composer.json [hunk 1]
```diff
@@ -25,8 +25,20 @@
         "doctrine/dbal": "^2.2"
     },
     "require-dev": {
+        "doctrine/common": "^2.7",
+        "doctrine/orm": "^2.5",
+        "doctrine/phpcr-odm": "^1.4 || ^2.0",
+        "jackalope/jackalope-doctrine-dbal": "^1.0",
+        "symfony/dependency-injection": "^3.4 || ^4.0",
```
Comment (jordisala1991) [inline composer.json]: why not 2.8 like in the other stable branches ?

File: src/Bridge/Symfony/Bundle/SonataDoctrineBundle.php [hunk 2]
```diff
@@ -0,0 +1,36 @@
+<?php
+
+declare(strict_types=1);
```
Comment (jordisala1991) [inline src/Bridge/Symfony/Bundle/SonataDoctrineBundle.php]: This needs to be php 5.6 compatible

File: tests/Adapter/AdapterChainTest.php [hunk 3]
```diff
@@ -0,0 +1,57 @@
+<?php
+
+declare(strict_types=1);
```
Comment (jordisala1991) [inline tests/Adapter/AdapterChainTest.php]: This class is not compatible with php 5.6

File: tests/Adapter/ORM/DoctrineORMAdapterTest.php [hunk 4]
```diff
@@ -0,0 +1,102 @@
+<?php
+
+declare(strict_types=1);
```
Comment (jordisala1991) [inline tests/Adapter/ORM/DoctrineORMAdapterTest.php]: This class is not compatible with php 5.6

File: tests/Adapter/PHPCR/DoctrinePHPCRAdapterTest.php [hunk 5]
```diff
@@ -0,0 +1,110 @@
+<?php
+
+declare(strict_types=1);
```
Comment (jordisala1991) [inline tests/Adapter/PHPCR/DoctrinePHPCRAdapterTest.php]: Same for this

**Comment by core23** [comment]
Please be aware, that i switched the namespaces:

 - Model/Adapter/AdapterChain => Adapter/AdapterChain
 - Model/Adapter/Doctrine*Adapter => Adapter/*/Doctrine*Adapter (* ORM / PHPCR)
 - DependenyInjection/ => Bridge/Symfony/DependencyInjection/
 - Model/Base*Manager => Model/*/Base*Manager (* MongoDB / ORM / PHPCR)

File: src/Model/BaseManager.php [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+                .' (http://symfony.com/doc/current/reference/configuration/doctrine.html#configuration-overview)'
+                .' or add the bundle to the `mappings` section in the doctrine configuration.',
+                $this->class
+            ));
+        }
+
+        return $manager;
+    }
+
+    public function getClass()
```
Comment (kunicmarko20) [inline src/Model/BaseManager.php]: No phpdoc comments at all on functions under, shouldn't comments be added?

Comment (core23) [reply src/Model/BaseManager.php]: Comment is here:

https://github.com/sonata-project/sonata-doctrine-extensions/pull/82/files#diff-be130aff7fc129752dd05b9b8b99a7eaR26

Comment (kunicmarko20) [reply src/Model/BaseManager.php]: Saw it after the comment, what about `inheritdoc` I think I saw somewhere we do not use that?

Comment (core23) [reply src/Model/BaseManager.php]: We don't use this anymore :(

https://github.com/sonata-project/dev-kit/issues/345

**Comment by greg0ire** [comment]
Just a sidenote: The classes are non-final, so we can extend this in the core bundle in with minor release
How about making them final anyway and using `class_alias`?

File: src/Model/MongoDB/BaseDocumentManager.php [hunk 7]
```diff
@@ -16,6 +16,8 @@
 
 /**
  * @author Hugo Briand <briand@ekino.com>
+ *
+ * @mixin DocumentManager
```
Comment (greg0ire) [inline src/Model/MongoDB/BaseDocumentManager.php]: Is that a standard phpdoc type?

**Comment by core23** [comment]
> How about making them final anyway and using `class_alias`?

Can you please elaborate

File: src/Model/BaseManager.php (hunk 6 above)
Comment (kunicmarko20) [reply src/Model/BaseManager.php]: that is it, I knew I saw it somewhere, ty for the link :smile:

File: src/Model/MongoDB/BaseDocumentManager.php (hunk 7 above)
Comment (core23) [reply src/Model/MongoDB/BaseDocumentManager.php]: Some IDE supports them and it doesn't hurt much ;)

**Comment by greg0ire** [comment]
Something like this:

```php
<?php
// Dependent.php

function test(Old $pwet)
{
}

test(new Newish);
```

```php
<?php

// Newish.php

final class Newish
{
}
```

```php
<?php

// Old.php

class_alias(Newish, Old);
```

See also: https://github.com/doctrine/doctrine2/pull/6886

**Comment by greg0ire** [comment]
It also means that method type hinting against old and receiving new will work *and* _vice versa_, which is not the case with extension.

**Comment by jordisala1991** [comment]
Keep in mind that old class must not be final otherwise this would be bc break

**Comment by greg0ire** [comment]
Ah so we won't be able to get `final` anyway. But better go with `class_alias` IMO.

**Comment by jordisala1991** [comment]
Yes class alias seems a great option anyway

**Comment by core23** [comment]
They are `final` now

**Comment by greg0ire** [comment]
> Keep in mind that old class must not be final otherwise this would be bc break

> Ah so we won't be able to get final anyway. But better go with class_alias IMO.

@core23 sorry, but I think you read too fast 😅

**Comment by core23** [comment]
Thought we can use `class_alias` im combination with your example code https://github.com/sonata-project/sonata-doctrine-extensions/pull/82#issuecomment-363195943?

So should I revert this?

**Comment by greg0ire** [comment]
Using `class_alias` 👍 
Using `final` 👎 (it's a BC-break), as pointed out by @jordisala1991

**Comment by core23** [comment]
Please review @kunicmarko20 @jordisala1991

File: src/Bridge/Symfony/Bundle/SonataDoctrineBundle.php [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+
+namespace Sonata\Doctrine\Bridge\Symfony\Bundle;
+
+use Sonata\Doctrine\Bridge\Symfony\DependencyInjection\SonataDoctrineExtension;
+use Symfony\Component\HttpKernel\Bundle\Bundle;
+
+final class SonataDoctrineBundle extends Bundle
+{
+    /**
+     * {@inheritdoc}
```
Comment (kunicmarko20) [inline src/Bridge/Symfony/Bundle/SonataDoctrineBundle.php]: Can you remove this?

File: src/Bridge/Symfony/Bundle/SonataDoctrineBundle.php [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+     */
+    public function getPath()
+    {
+        return __DIR__.'/..';
+    }
+
+    /**
+     * {@inheritdoc}
```
Comment (kunicmarko20) [inline src/Bridge/Symfony/Bundle/SonataDoctrineBundle.php]: This one too

File: src/Bridge/Symfony/DependencyInjection/SonataDoctrineExtension.php [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+ * @author Thomas Rabaix <thomas.rabaix@sonata-project.org>
+ */
+class SonataDoctrineExtension extends Extension
+{
+    public function load(array $configs, ContainerBuilder $container)
+    {
+        $loader = new XmlFileLoader($container, new FileLocator(__DIR__.'/../Resources/config'));
+        $loader->load('services.xml');
+
+        if (class_exists(Doctrine\ORM\EntityManagerInterface::class)) {
```
Comment (kunicmarko20) [inline src/Bridge/Symfony/DependencyInjection/SonataDoctrineExtension.php]: Is `::class` even needed here? Or can you replace the FQCN with `use`?

File: src/Bridge/Symfony/DependencyInjection/SonataDoctrineExtension.php [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+            $loader->load('doctrine_orm.xml');
+        }
+
+        if (class_exists(Doctrine\ODM\PHPCR\DocumentManager::class)) {
```
Comment (kunicmarko20) [inline src/Bridge/Symfony/DependencyInjection/SonataDoctrineExtension.php]: Same here

File: src/Bridge/Symfony/Resources/config/doctrine_orm.xml [hunk 12]
```diff
@@ -0,0 +1,8 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<container xmlns="http://symfony.com/schema/dic/services" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://symfony.com/schema/dic/services http://symfony.com/schema/dic/services/services-1.0.xsd">
+    <services>
+        <service id="sonata.doctrine.adapter.doctrine_orm" class="Sonata\Doctrine\Adapter\ORM\DoctrineORMAdapter" public="false">
```
Comment (kunicmarko20) [inline src/Bridge/Symfony/Resources/config/doctrine_orm.xml]: Isn't public by default `false`?

File: src/Bridge/Symfony/Resources/config/doctrine_phpcr.xml [hunk 13]
```diff
@@ -0,0 +1,8 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<container xmlns="http://symfony.com/schema/dic/services" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://symfony.com/schema/dic/services http://symfony.com/schema/dic/services/services-1.0.xsd">
+    <services>
+        <service id="sonata.doctrine.adapter.doctrine_phpcr" class="Sonata\Doctrine\Adapter\PHPCR\DoctrinePHPCRAdapter" public="false">
```
Comment (kunicmarko20) [inline src/Bridge/Symfony/Resources/config/doctrine_phpcr.xml]: Same here

File: src/Document/BasePHPCRManager.php [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+    {
+        if ('dm' === $name) {
+            return $this->getObjectManager();
+        }
+
+        throw new \RuntimeException(sprintf('The property %s does not exists', $name));
+    }
+
+    /**
+     * {@inheritdoc}
```
Comment (kunicmarko20) [inline src/Document/BasePHPCRManager.php]: In this case, do we still remove `inheritdoc`?

File: src/Document/BasePHPCRManager.php [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+     *
+     * @throws \LogicException Each call
+     */
+    public function getConnection()
+    {
+        throw new \LogicException('PHPCR does not use a database connection.');
+    }
+
+    /**
+     * {@inheritdoc}
```
Comment (kunicmarko20) [inline src/Document/BasePHPCRManager.php]: Same here

File: src/Model/BaseManager.php [hunk 16]
```diff
@@ -0,0 +1,142 @@
...
+    /**
+     * @return ObjectManager
+     */
+    public function getObjectManager()
+    {
+        $manager = $this->registry->getManagerForClass($this->class);
+
+        if (!$manager) {
+            throw new \RuntimeException(sprintf(
+                'Unable to find the mapping information for the class %s.'
```
Comment (kunicmarko20) [inline src/Model/BaseManager.php]: Shouldn't the comment have `@throws`

File: tests/Document/BaseDocumentManagerTest.php [hunk 17]
```diff
@@ -0,0 +1,37 @@
...
+{
+}
+
+final class BaseDocumentManagerTest extends TestCase
+{
+    public function getManager()
+    {
+        $registry = $this->createMock(ManagerRegistry::class);
+
+        $manager = new DocumentManager('classname', $registry);
```
Comment (kunicmarko20) [inline tests/Document/BaseDocumentManagerTest.php]: Can't you just return here?

File: tests/Entity/BaseEntityManagerTest.php [hunk 18]
```diff
@@ -0,0 +1,70 @@
...
+{
+}
+
+final class BaseEntityManagerTest extends TestCase
+{
+    public function getManager()
+    {
+        $registry = $this->createMock(ManagerRegistry::class);
+
+        $manager = new EntityManager('classname', $registry);
```
Comment (kunicmarko20) [inline tests/Entity/BaseEntityManagerTest.php]: Same here

**Comment by kunicmarko20** [comment]
I am done, looks 👌, just a few comments.

File: src/Bridge/Symfony/Resources/config/doctrine_orm.xml (hunk 12 above)
Comment (core23) [reply src/Bridge/Symfony/Resources/config/doctrine_orm.xml]: Only in symfony 4

File: src/Document/BasePHPCRManager.php (hunk 14 above)
Comment (core23) [reply src/Document/BasePHPCRManager.php]: We shouldn't do this, cause the original documentation will be lost if you generate api docs

Comment (greg0ire) [reply src/Document/BasePHPCRManager.php]: Do people actually do this?

Comment (core23) [reply src/Document/BasePHPCRManager.php]: That's how PHPDoc works...

**Comment by greg0ire** [comment]
Thanks @core23 !

**Comment by core23** [comment]
First step done. 

Next step: We need a release @Soullivaneuh !

**Pull request merged**